Middleware for exeptions. Pattern Fasade for repository. Linq for FileParser.

AndreyErin opened the pull request:
Создано middleware для глобальной обработки ошибок.
Методы FileParser переписаны на LINQ.
FileValidator теперь возвращает сообщение об ошибке.
Создан статический класс Config, для чтения данных из конфигурационного файла(пути к json-репозиториям).

DAL\Repositories\Base\Repository.cs - Является фасадом над над классами:
DAL\FileAccess\
FileRepository.cs
RepositoryReader.cs
RepositoryWriter.cs

RepositoryReaderExtensions.cs содержит метод TryReadData, для проверки возможности (и)чтения из репозитория(json-файл)

AndreyErin committed:
Борьба с излишними вложенностями в классах AdvertisingPlatfirmService и FileRepository(вынос блоков кода в отдельные функции)

AndreyErin committed:
Создан LocationsService теперь AdvertisingPlatformsService работает с сервисом локация, а не с репозиторием локаций. Метод обновления данных сервиса вынесен в отдельный интерфейс IReplaceData. Мелкие правки

AndreyErin committed:
Созданны сервисы парсинга и валидации. Json-файлы теперь сохраняют данные не в одну строчку, а в виде структуры

AndreyErin committed:
Json-файлы теперь сохраняют данные не в одну строчку, а в виде структуры

AndreyErin committed:
Переименование папки с  json-файлами

AndreyErin committed:
Создан ExeptionHandlerMiddleware для глобальной обработки ошибок. Добавлены кастомные исключения быбрасываемые из FileReader'a: InvalidFileDataExeption и invalidFileExeption. Добавлены xml-комментарии.

AndreyErin committed:
Добавлен класс Config для чтения файла конфигурации и пердоставления путей к json-фалам посредствам статических свойств.  Добавлено ConfigureReadExeption,  удалено InvalidFileExeption. Мелкие правки.

AndreyErin committed:
мелкие правки

AndreyErin committed:
Мелкие правки

AndreyErin committed:
Добавление проекта Business и перенос в него сервисов, абстрактного класса FileRepository (и папки с методами расширения к нему). Настройка зависимостей

AndreyErin committed:
Переработка методов парсера

AndreyErin committed:
Тескт вынесен в файлы ресурсов, для доступа к файлам ресурсов создан статический класс Message. Добавлена модель MessageOfReplace для ответа, в случае успешном обновлении данных.

AndreyErin committed:
Добавлена модель ResponseModel, для возврата результата из контроллера. Добавлен Swagger

AndreyErin committed:
Добавлены модели ответа из котроллера, добавлен Swagger, добавлены xml-описания моделей

AndreyErin committed:
Добавление xml-комментариев,  мелкие правки

AndreyErin committed:
Свойства модели  AdvertisingInformation теперь относятся к интерфейсу IReadOnlyList. Поправлены классы работающие с этой моделью.

chaika-ds committed:
Merge pull request #5 from AdvPlatforms/module_01a

Add Validator and Parcer for FileReader. Add LocationService. Refactoring parce methods.

AndreyErin committed:
Модель ResponseApi переименована в BaseResponce и сделана абстрактным классом. ApiResultData удален. Комментарии в контроллере о переносе кода в будующем помечены TODO:. Метод FileValidator.IsValid() переименован в IsValidAdvertisingData. Мелкие правки.

AndreyErin committed:
Слияние module_01 и module_02

AndreyErin committed:
Мелкие правки для востановления работоспособности программы после слияния

AndreyErin committed:
Папка Resources и класс FileRepository вынесены в DAL. FileRepository теперь не абстрактный класс. FileRepositoriesExtentions разделен на 2 класса RepositoryReader и RepositoryWriter. Создан класс Repository являющийся фасадом над классами FileRepository, RepositoryReader и RepositoryWriter. Config из сервиса переделан в класс со статическими полями которые инициализируются в файле Program путем создания экземпляра Config.

AndreyErin committed:
Метод FileParcer.GetLocations() переписан на LINQ.

AndreyErin committed:
Остальные методы FileParser переписаны на LINQ.

AndreyErin committed:
Config переделан в статический класс. Перенос классов и папок между проектами. Мелкие правки.

AndreyErin committed:
Добавлена генерация исключений в классы: FileRepository, RepositoryReader, RepositoryWriter.

AndreyErin committed:
Для RepositoryReader добавлено расширение TryReadData, где проверяется возможность получить данные из репозитория и в случае успеха параметром возвращаются это данные.

AndreyErin committed:
AdvertisingPlatformsController теперь выбрасывает исключени я в случае неудачи. Добавлены классы исключений, xml-комментарии, мелкие правки.

AndreyErin committed:
Доработан FileValidator(сообщает разные сообщения об ошибках). FileParser теперь ищет данные по шаблону, что отбрасывает некорректную информацию. Мелкие правки.

KhristinaVorozhbet93 commented on AdvertisingPlatforms.Domain/Middlewares/ExeptionHanglerMiddleware.cs:
Middleware по глобальной обработке ошибок стоит перенести в слой WebApi.
Бизнес-слой может генерировать исключения, которые затем обрабатываются слоем API, но преобразованием ошибок в HTTP ответы должен заниматься слой API)

KhristinaVorozhbet93 commented on AdvertisingPlatforms.Domain/Configuration/Config.cs:
В данном случае конфигурации не должны находится в слое Domain, слой Domain ничего не должен знать про IConfiguration, нужно вынести конфигурации в слой WebApi, там считывать с appsetting необходимые данные, а далее внедрять при помощи, например, IOptions в необходимые сервисы)

AndreyErin replied:
Config переименован в DbConfig и перенесен в проект DAL(т.к. только в нем он используется)

KhristinaVorozhbet93 commented on AdvertisingPlatforms.Business/Middlewares/ExeptionHanglerMiddleware.cs:
Middleware по глобальной обработке ошибок стоит перенести в слой WebApi.
Бизнес-слой может генерировать исключения, которые затем обрабатываются слоем API, но преобразованием ошибок в HTTP ответы должен заниматься слой API)

AndreyErin replied:
Сделано)

AndreyErin committed:
Перенос папки Middlewares из проекта Bussines в проект WebApi. Config переименован в DbConfig и перенесен в проект DAL.

KhristinaVorozhbet93 commented on AdvertisingPlatforms.Domain/Models/BaseModels/Resource.cs:
Смотри, вот у тебя есть тип Resource и есть обобщенный репозиторий  public class Repository<TResource> where TResource: Resource, где ты указываешь что тип обязательно должен быть Resource, но вот например метод DeleteFromRepository(int id) на вход принимает id, а сам твой класс Resource не обязует классы-наследники чтобы поле id было обязательным, и если его не сделать обязательным, то мы без проблем, например, сможем создать объект класса AdvertisingPlatform без идентификатора. Стоило бы лучше сделать поле int abstract и обязать классов-наследников его реализовывать. А в самом классе-наследнике сделать инкапсуляцию, и конструктор, чтобы это поле обязательно указывалось.

AndreyErin replied:
Сделано

KhristinaVorozhbet93 commented on AdvertisingPlatforms.DAL/Repositories/Base/Repository.cs:
Вообще, если делать обобщенный репозиторий, то я бы сделала тут только базовые методы CRUD, поскольку они присущи в основном ко всему, а методы получения по имени можно было бы выделить в отдельный конкретный репозиторий, унаследованный об базового, потому как не у каждой сущности может быть имя, а вот идентификатор в основном.  Если всегда добавлять методы в базовый репозиторий, то он может разрастись и одному сервису эти методы будут нужны, а другому - нет - избыточность)

AndreyErin replied:
Тут была задача сделать фасад над 3мя классами: FileRepository, RepositoryWriter, RepostitoryReader. Отсюда и лишние методы.

chaika-ds commented on AdvertisingPlatforms.Business/Services/AdvertisingServices/AdvertisingPlatformsService.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            try
+            {
+                List<string> result = new();
```
Лучше возвращать параметр в методе ка результат выполненных операций, подумай как можно переработать метод, чтобы не объявлять отдельно result, а возвращать его после рассчета

AndreyErin replied:
результат теперь возвращается как результат операции, либо null

chaika-ds commented on AdvertisingPlatforms.DAL/Resources/Messages.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <summary>
+        /// Messages for errors.
+        /// </summary>
+        public static class Error
```
1. Класс в классе не самая лучшая архитектура, желательно разносить для читаемости
2. Выносить в ресурс нужно или например json для тестовых целей или настройки внутренней конфигурации проекта. Описание ошибок лучше выделить в папку Const и разбить по классам сущностей 
3. Описывай ошибки на английском, лаконичнее и универсальнее

AndreyErin replied:
Сделано

AndreyErin replied:
Из Message выделено 3 класса констант. Сам класс Message переименован в RuLocalization и теперь содержит метод для локализации констант. Сейчас локализация используется для  вывода ошибки в браузер(в промежуточном ПО по обработке ошибок)

chaika-ds commented on AdvertisingPlatforms.DAL/Resources/Messages.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <summary>
+        /// Messages for information.
+        /// </summary>
+        public static class Information 
```
Это тоже в отдельный класс констант и вынести, сделать классом

AndreyErin replied:
Сделано

chaika-ds commented on AdvertisingPlatforms.Business/Services/AdvertisingServices/LocationsService.cs:
```diff
@@ -0,0 +1,59 @@
+
+                return newEntitiesList.Count;
+            }
+            catch (Exception ex)
```
Несоответствие отлавливаемой ошибки возвращаемой плюс  throw new LocationsServiceExeption перебивает весь стек
Либо вынести отлов ошибок в контроллер и нормализировать, чтобы при ошибке спланировать запись в логи и возвращать например 500 в ответе контроллера, либо добавить ошибки в Middleware.

AndreyErin replied:
Middleware теперь обрабатывает стектрейс внутренних исключений. Все виды кастомных исключений изменены на единственное - BusinessException возвращающее 400.

chaika-ds commented on AdvertisingPlatforms.Business/Services/AdvertisingServices/LocationsService.cs:
```diff
@@ -0,0 +1,59 @@
+            {
+                return _locationRepository.GetByNameFromRepository(name);
+            }
+            catch (Exception ex)
```
Вынести обработку ошибки в контроллер или Middleware

AndreyErin replied:
Все виды кастомных исключений изменены на единственное - BusinessException возвращающее 400.

chaika-ds replied:
В общем по всей обработке ошибок переносим на следующий модуль

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileParser.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <returns>Data.</returns>
+        public AdvertisingInformation GetParseData(string fileContent)
+        {
+            var regex = new Regex(@"^[А-Яа-я.\- ]+:[A-Za-z,\/]+$");
```
Желательно вынести Regex из метода, возможно в константы или хотябы параметры класса

AndreyErin replied:
Паттерн вынесен в константы, сам объект теперь является приватным полем класса.

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileParser.cs:
```diff
@@ -0,0 +1,85 @@
+            var regex = new Regex(@"^[А-Яа-я.\- ]+:[A-Za-z,\/]+$");
+
+            var advertisingPlatformsAndLocationRaw = fileContent
+                .Split("\r\n")
```
Тут строковые переменные тоже вынести в контсанты

AndreyErin replied:
Сделано

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileParser.cs:
```diff
@@ -0,0 +1,85 @@
+        {
+            var regex = new Regex(@"^[А-Яа-я.\- ]+:[A-Za-z,\/]+$");
+
+            var advertisingPlatformsAndLocationRaw = fileContent
```
Может тут выделить в отдельный метод совместный парсинг локаций и реклам

AndreyErin replied:
Сделано

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileParser.cs:
```diff
@@ -0,0 +1,85 @@
+                .ToDictionary();
+
+
+            var locations = GetLocations(locationsRaw, advertisingPlatforms);
```
Тут не совсем понял, сначала рекламу и локации разделем, потом соединяем, давай обсудим логику этих преобразований

AndreyErin replied:
Добавлена третья модель AdvertisingInLocation содержащая отношения реклам к локациям.

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileParser.cs:
```diff
@@ -0,0 +1,85 @@
+                })
+                .ToList();
+
+            locations.ForEach(x=> x.AdvertisingPlatformIds.AddRange(GetAdditionalIds(locationsRaw, advertisingPlatforms, x.Name)));    
```
Можно внести логику ForEach в Select до преобразования в ToList

AndreyErin replied:
ForEach заменен на дополнительную секцию Select. Метод теперь возвращает IEnumerable<>

AndreyErin replied:
FileParcer очередной раз переработан) Приватные методы в любом случае возвращают IEnumerable<>

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileParser.cs:
```diff
@@ -0,0 +1,85 @@
+            return locationsRaw.Where(x => currentAdvertisingPlatorm.Contains(x.Key) && x.Key != currentAdvertisingPlatorm)
+                               .Select(x => x.Value)
+                               .Join(advertisingPlatforms, x => x, y => y.Name, (x, y) => y.Id)
+                               .ToList();
```
Тут можно и без преобразования в ToList, все равно дальше преобразуешь, заодно будет видно отложенное исполнение

AndreyErin replied:
метод теперь возвращает IEnumerable<>

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileReader.cs:
```diff
@@ -0,0 +1,53 @@
+
+            var isValid = _validator.IsValidAdvertisingData(fileContent);
+
+            if (!isValid.result)
```
Тут в следующих модулях добавим библиотеку с валидатором, но пока ловится InvalidFileDataExeption, которая нигде не обрабатывается

AndreyErin replied:
Все виды кастомных исключений изменены на единственное - BusinessException. Обрабатывается в middleware и возвращает 400.

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileValidator.cs:
```diff
@@ -0,0 +1,37 @@
+    /// </summary>
+    public class FileValidator : IFileValidator
+    {
+        const string Splitter = ":";
```
Вынести в общие константы

AndreyErin replied:
Сделано

chaika-ds commented on AdvertisingPlatforms.Business/Services/FileHandlingServices/FileValidator.cs:
```diff
@@ -0,0 +1,37 @@
+                return (false, Messages.Error.FileHaveShortData);
+
+            if(Regex.Matches(data, @"\r\n").Count == 0 &&
+               !Regex.IsMatch(data, @"^[А-Яа-я.\- ]+:[A-Za-z,\/]+$"))
```
Строкив константы

AndreyErin replied:
Сделано

chaika-ds commented on AdvertisingPlatforms.DAL/Configuration/DbConfig.cs:
```diff
@@ -0,0 +1,40 @@
+        /// Initialize configuration.
+        /// </summary>
+        /// <param name="configuration">Configuration.</param>
+        public static void Initialize(IConfiguration configuration)
```
Было бы интересно сделать LazyLoad инициализацию при необходимости обращения к бд и не тянуть ее в класс Program

AndreyErin replied:
Для инициализации DbConfig создан Host-сервис ConfigInitializer. Поля DbConfig теперь имеют тип Lazy<string>.

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/Extensions/RepositoryReaderExtensions.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <returns></returns>
+        public static bool TryReadData<TResource>(this RepositoryReader repositoryReader, string filePath,[NotNullWhen(true)] out List<TResource>? data) where TResource : Resource
+        {
+            data = null;
```
1. Зачем отдельно объвлять переменную и использовать затем только для получения данных, лучше var data = JsonSerializer...
2. Для чего здесь в параметрах repositoryReader?

AndreyErin replied:
1. data  тут выходной параметр. и если не инициализировать ее в начале, то в случае неудачи придется писать data = null в 2х местах(перед каждым return false;)
2. repositoryReader заменен на _

chaika-ds replied:
Убери repositoryReader, с гим тут ничего не делается. Метод называтся TryReadData, значит ты делаешь манипуляции с путем файла, с которого пытаешьсяя считать, соответственно его в this и ставь.

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/Extensions/RepositoryReaderExtensions.cs:
```diff
@@ -0,0 +1,39 @@
+
+                return data != null;
+            }
+            catch (Exception)
```
Тут может указать что если ошибка в серриализации то возвращаем false, а если глобально шото падает, то ловить дальше по коду.
Плюс если входящие даныне не равны null, то и выходящие не должны быть нулевыми, не вижу смысл в проверке data != null

AndreyErin replied:
Теперь ошибки отлавливаются только при десериализации.
Конструкция data != null нужна для работы с атрибутом [NotNullWhen(true)]. Атрибут в свою очередь нужен, чтобы сообщить компилятору в вызывающем коде, что в случае возврата true не нужно подозревать data в том, что она может быть равна null.

chaika-ds replied:
Тогда нужна обработка варианта при котором data = null

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/FileRepository.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <param name="repositoryWriter">Writer for repository.</param>
+        public void AddToRepository(TResource entity, IRepositoryReader repositoryReader, IRepositoryWriter repositoryWriter)
+        {
+            var db = repositoryReader.GetAllFromFile<TResource>(_filePath);
```
Это точно не db, тут просится что то вроди entityRepositoryData

AndreyErin replied:
Переименовано в entities

AndreyErin replied:
Еще раз переименовано в repositoryEntities

chaika-ds replied:
Сделай FilePath приватным, ты его только в этом классе используешь

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/FileRepository.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <param name="repositoryWriter">Writer for repository.</param>
+        public void DeleteFromRepository(int id, IRepositoryReader repositoryReader, IRepositoryWriter repositoryWriter)
+        {
+            var db = repositoryReader.GetAllFromFile<TResource>(_filePath);
```
переименуй db и дальше по классу

AndreyErin replied:
Переименовано в entities

AndreyErin replied:
Еще раз переименовано в repositoryEntities

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/RepositoryReader.cs:
```diff
@@ -0,0 +1,30 @@
+            }
+            else
+            {
+                throw new ReadRepositoryExeption(Messages.Error.ReadRepository);
```
Нет обработки ошибки при ее получении

AndreyErin replied:
Все виды кастомных исключений изменены на единственное - BusinessException. Обрабатывается в middleware и возвращает 400.

chaika-ds replied:
Если делаешь кастомные исключения, то и обрабатывать каждый тип исключения нужно сообветственно. Как вариант расширить количество обрабатываемых исключений в  middleware. Иначе будет путать что все сводится к BusinessException

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/RepositoryWriter.cs:
```diff
@@ -0,0 +1,40 @@
+        {
+            try
+            {
+                if (string.IsNullOrEmpty(filePath))
```
Это можно в валидатор выделить

AndreyErin replied:
Проверка входящих аргументов убрана, т.к. в данном случае она повторяет функционал try-catch внутри которой находится. Выбрасывает тот же самый ArgumentException.

chaika-ds commented on AdvertisingPlatforms.DAL/FileAccess/RepositoryWriter.cs:
```diff
@@ -0,0 +1,40 @@
+                sw.Write(newDbJson);
+
+            }
+            catch (Exception ex)
```
Тут тоже ломается stacktrace при обработке ошибок ((

AndreyErin replied:
Middleware теперь обрабатывает стектрейс внутренних исключений.

chaika-ds commented on AdvertisingPlatforms.DAL/Repositories/Base/Repository.cs:
```diff
@@ -0,0 +1,102 @@
+        private readonly IRepositoryWriter _repositoryWriter;
+        public Repository()
+        {
+            var filePath = String.Empty;
```
Вынести логику получения filePath  из конструктора и сделать var filePath = switch...

AndreyErin replied:
Логика вынесена в отдельный метод GetFilePath(). В методе реализован switch возвращающий результат.

chaika-ds commented on AdvertisingPlatforms.DAL/Repositories/Base/Repository.cs:
```diff
@@ -0,0 +1,102 @@
+    /// <typeparam name="TResource">Resource.</typeparam>
+    public class Repository<TResource> where TResource: Resource
+    {
+        private readonly FileRepository<TResource> _repository;
```
Если уж сделал интерфейс, отправь его в DI и сделай агрегацию

AndreyErin replied:
RepositoryReader и RepositoryWriter теперь получаются через DI.

chaika-ds commented on AdvertisingPlatforms/Middlewares/ExeptionHanglerMiddleware.cs:
```diff
@@ -0,0 +1,60 @@
+            var message = exception.Message;
+            var statusCode = 0;
+
+            switch (exception)
```
Добавил пример из одного из моих предыдущих проектов, там можно расширить количество типов отлавливаемых ошибок, сбор ошибок по stacktrace и красиво их возвращать
[Exception message.txt](https://github.com/user-attachments/files/20129702/Exception.message.txt)

AndreyErin replied:
Перетащил себе идеи: 
1) Показывать детали только если это среда разработки.
2) Отображать в деталях информацию о внутренних ошибках (стектрейс, тип, сообщение)
3) Модель для информации об ошибках.

Количество отлавливаемых ошибок наоборот уменьшилось до 1й))
Теперь в middleware может поймать либо BusinessException (с разным описанием и возможно внутренним исключением) и вернуть 400.
Либо поймать любую непредвиденную и показать 500.

chaika-ds commented on AdvertisingPlatforms/ServiceCollection/ServiceExtensions.cs:
```diff
@@ -1,21 +1,66 @@
-            services.AddScoped<FileRepository<AdvertisingPlatform>, AdvertisingPlatformsFileRepository>();
+            services.AddScoped<ILocationsService, LocationsService>();
+
+            return services;
```
Зачем возвращать services во всех расширениях, если это не используется, лучше мотоды сделать void

AndreyErin replied:
Сделано

Vadya0994 commented on AdvertisingPlatforms.DAL/FileAccess/FileRepository.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <param name="name">name of entity.</param>
+        /// <param name="repositoryReader">Reader for repository.</param>
+        /// <returns>Entity for success, null for fail</returns>
+        public TResource? GetByNameFromRepository(string name, IRepositoryReader repositoryReader)
```
что-то Repository лишнее в названии методов, у тебя класс говорит сам за себя, что это Repository

AndreyErin replied:
Было добавлено по замечанию прошлого модуля.
https://github.com/AdvPlatforms/AdvertisingPlatforms/pull/3#discussion_r2039860275

Vadya0994 commented on AdvertisingPlatforms/Middlewares/ExeptionHanglerMiddleware.cs:
```diff
@@ -0,0 +1,60 @@
+                        statusCode = 422;
+                    break;
+                case EntityNotFoundExeption:
+                    statusCode = 404;
```
Что-то перемудрил с кодами ошибок и с кастомными исключениями. Просто создай один класс BusinessException, которым будешь бросаться и тут его лови, и возвращай просто 400 ошибку.

AndreyErin replied:
Сделано

Vadya0994 commented on AdvertisingPlatforms/Program.cs:
```diff
@@ -8,25 +10,42 @@ public static void Main(string[] args)
             var builder = WebApplication.CreateBuilder(args);
             builder.Services.AddControllers();          
+            
+            builder.Services.AddAdvertisingServices();
```
Конфигурацию DI можно спрятать в отдельный класс

AndreyErin replied:
Настройки  builder спрятаны в метод расширения ConfigureBuilder.

Vadya0994 commented on AdvertisingPlatforms/Program.cs:
```diff
@@ -8,25 +10,42 @@ public static void Main(string[] args)
-            builder.Services.AddServices();
+            if (builder.Environment.IsDevelopment())
+            {
+                builder.Services.AddSwaggerServices();
```
это по сути тоже относится к конфигурации DI

AndreyErin replied:
Настройки  builder спрятаны в метод расширения ConfigureBuilder.

Vadya0994 commented on AdvertisingPlatforms/Program.cs:
```diff
@@ -8,25 +10,42 @@ public static void Main(string[] args)
 
+            if (app.Environment.IsDevelopment())
+            {
+                app.UseSwagger();
```
Конфигурацию Middleware (все что с приставкой use) тоже можно спрятать в отдельный класс, туда же что и конфигурацию DI.

AndreyErin replied:
Настройки  app спрятаны в метод расширения ConfigureApp.

Vadya0994 commented on AdvertisingPlatforms/ServiceCollection/ServiceExtensions.cs:
```diff
@@ -1,21 +1,66 @@
 
 namespace AdvertisingPlatforms.ServiceCollection
 {
     public static class ServiceExtensions
```
методы этого класса корректнее попрятать в сборки, которые данные методы подключают.

AndreyErin replied:
Сделано

Vadya0994 commented on AdvertisingPlatforms.Business/Services/AdvertisingServices/AdvertisingPlatformsService.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <param name="locationName">Name of location.</param>
+        /// <returns>Advertising platform names.</returns>
+        /// <exception cref="AdvertisingPlatformsServiceExeption"></exception>
+        public List<string> GetAdvertisingPlatformsForLocation(string locationName)
```
Я бы в открытых методах не возвращал бы List, потенциально из вне даешь возможность его модифицировать, ReadOnlyList реально ли вернуть?

Vadya0994 replied:
Ну и так везде в других местах

AndreyErin replied:
Сделано

Vadya0994 commented on AdvertisingPlatforms.DAL/Resources/Messages.cs:
```diff
@@ -0,0 +1,53 @@
+    /// <summary>
+    /// Class for working with message resources.
+    /// </summary>
+    public static class Messages
```
Я бы назвал это класс LocalizationConstants, и хранил бы только строковые константы, а ResourseManager бы вынес в отдельный статический класс c реализацией метода получения локализованной строки по переданной константе.

AndreyErin replied:
Из Message выделено 3 класса констант. Сам класс Message переименован в RuLocalization и теперь содержит метод для локализации констант.

AndreyErin committed:
AdvertisingPlatform  и Location теперь обязуются инициализировать id. AdvertisingPlatformService.GetAdvertisingPlatformsForLocation() теперь возвращает IOnliList, из этого же метода убрана промежуточная переменная "result".

AndreyErin committed:
В методе расширения TryReadData() для FileRepository имя параметра отвечающее за класс самого репозитория было отброшено(заменено на _), т.к. не использовалось в самомо методе. Try-catch теперь перехватывает ошибки связанные только с десериализацией.

AndreyErin committed:
мелкие правки

AndreyErin committed:
Наследники класса Repository теперь получают RepositoryReader и RepositoryWriter через DI. Логика из конструктора Repostitory вынесена в отдельный метод GetFilePath.

AndreyErin committed:
Строки из FileParser и FileValidator вынесены в класс констант. FileParser теперь возвращает листы только в самом конце, остальные методы возвращают IEnumerable<>

AndreyErin committed:
Из Message выделено 3 класса констант. Сам класс Message переименован в RuLocalization и теперь содержит метод для локализации констант.

AndreyErin committed:
Создан класс BusinessException, остальные удалены. Создана модель информации об ошибке ExceptionInfo. В middlevare по обработке ошибок добавлен метод проверки среды разработки, а также обработка InnerException. Мелкие правки.

AndreyErin committed:
Из Program в методы расширения вынесены настройки для WebApplicationBuilder и WebAplication. Методы класса ServiceExtensions разнесены по соответсвующим сборкам.

AndreyErin committed:
ExceptionHandlerMiddleware теперь озвращает ответ в виде json. Из метода FileParser.GetParseData() выделен отдельный метод SplitIntoPlatformAndLocationsRaw(). Мелкие правки

AndreyErin committed:
Для инициализации DbConfig создан Host-сервис ConfigInitializer. Поля DbConfig теперь имеют тип Lazy<string>. Мелкие правки.

AndreyErin committed:
DbConfig теперь выбрасывает кастомное исключение при попытке обратиться к полю в случае если DbConfig не был инициализирован. Middleware для обработки ошибок теперь возвращает описание ошибок на английском. Функционал по локализации констант больше не задействован.

AndreyErin committed:
Добавлена новая модель AdvertisingsInLocation содержащая в себе список рекламных площадок для конкретной локации. Добавлен соответвующий сервис. Модель Location больше не осодержит списка рекламных площадок, изменен соответвующий сервис.

AndreyErin committed:
Добавлен репозиторий для моделей AdvertisingInLocation. Метод FileRepository.GetByName удален. Из базовой модели Resource удалено свойство Name. Мелкие правки.

AndreyErin committed:
Метод GetByName() сделан расширением для Repostitory<Location> т.к. на данный момент этот метод нужен только для поиска локаций.

AndreyErin committed:
ExceptionHandlerMiddleware теперь возвращает информацию о внутреннем исключении  в виде объекта. Мелкие правки.

AndreyErin committed:
Мелкие правки.

AndreyErin committed:
Сделал filePath снова приватным.
Вернул GetByNameFromRepository в базовый класс FileRepository. Вернул name в базовую модель Resource. Мелкие правки.
